### PR TITLE
Adding Org & Space Name Annotations

### PR DESCRIPTION
**Pull request by mkkeffeler**
### What does this PR do?
Discussed as part of https://github.com/cyberark/conjur-service-broker/issues/238, We are using the service broker 2.15 API Spec to add ```organization_name``` and ```space_name``` as annotations, when they exist.

### What ticket does this PR close?
Resolves https://github.com/cyberark/conjur-service-broker/issues/238

#### Change log
- [ X] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [X ] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ X] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

### CODE REVIEW

**Comment by mkkeffeler** [comment]
To whomever is reviewing, I had a hard time making sure I had run the tests sufficiently. I ran the unit tests (don't have dependencies needed to run E2E) and got this.

```
SC-MKEFFEL-OSX:conjur-service-broker mkeffele$ ./dev/test_unit
++ dirname ./dev/test_unit
+ cd ./dev
++ pwd
+ docker run -e CONJUR_ACCOUNT=cucumber -e CONJUR_AUTHN_LOGIN=admin -e SECURITY_USER_NAME=TEST_USER_NAME -e SECURITY_USER_PASSWORD=TEST_USER_PASSWORD -v /Users/mkeffele/git_repos/conjur-service-broker/dev/..:/app conjur-service-broker-test bundle exec rspec --format RspecJunitFormatter --out spec/reports/test.xml
SC-MKEFFEL-OSX:conjur-service-broker mkeffele$ 
```

There was another time I ran and got a syntax error, so I know it is running the tests, but since there wasn't a clean output I can't tell you for sure it worked as expected.

**Comment by mkkeffeler** [comment]
Wou

> ### What does this PR do?
> Discussed as part of #238, We are using the service broker 2.15 API Spec to add `organization_name` and `space_name` as annotations, when they exist.
> 
> ### What ticket does this PR close?
> Resolves #238
> 
> ### Checklists
> #### Change log
> * [ ]  The CHANGELOG has been updated, or
> * [ ]  This PR does not include user-facing changes and doesn't require a CHANGELOG update
> 
> #### Test coverage
> * [ ]  This PR includes new unit and integration tests to go with the code changes, or
> * [ ]  The changes in this PR do not require tests
> 
> #### Documentation
> * [ ]  Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
> * [ ]  This PR does not require updating any documentation

**Comment by mkkeffeler** [comment]
Also, It may make sense for the annotations to be ```pcf/orgName``` but would want guidance on the teams thoughts.

**Comment by izgeri** [comment]
Thanks @mkkeffeler! I noted on the issue that we'll take a look at this tomorrow - at a high level it looks great, but we'll let you know if we have any feedback on the tests / the specific annotations, etc.

**Comment by mkkeffeler** [comment]
I thought on this a little more. 

Another thing people often want is to know what the layer is representing, whether it's a space or an org. Since we include org name on the space (which is needed in my opinion) we can't use those annotations to provide that info (because the org name annotation is on both spaces and orgs) 

we need another annotation like "type" with a value of "org" or "space". If we add that people can easily filter on the list of orgs and list of spaces when trying to grant authorization.

**Review by diverdane — COMMENTED** [review]
@mkkeffeler,

Thank you for submitting this PR! This looks very good!
I've added some comments and suggestions.

One thing that might be good to add is an scenario/feature test that sets up values for the organization_name and space_name fields in the Service Broker API context. This would be VERY easy to add... it would be a minor variation of this test:
https://github.com/cyberark/conjur-service-broker/blob/main/features/provision.feature#L48-L66

Also, for locally running unit tests as well as the feature/scenario tests, we have a pretty cool development environment. It's all menu-driven, VERY easy to use, and allows you to change code on the fly and re-test. It's described in the CONTRIBUTING.md here:
https://github.com/cyberark/conjur-service-broker/blob/main/CONTRIBUTING.md#development-environment

As it says in the instructions, you just need to run 3 commands to get started:
```
./dev/build
./dev/build_dev
./dev/start
```
and then select from the menu.  The scripts will set up some local Conjur instances running in Docker-Compose, and allow you to test using those instances (esp. useful if you add a feature/scenario test).

Thanks again for the contribution!
-Dane

**Comment by diverdane** [comment]
I forgot to mention, when you're using the menu-driven dev environment to run tests locally, you can view local Conjur server logs to help troubleshoot any failing tests. The scripts create a bunch of local Conjur servers, one for "happy path" testing and the rest for testing invalid configurations. You can tail the "happy path" server by running the following in a separate window:
```
docker logs -f conjur-service-broker_conjur-service-broker_1
```

is the one using port 3000,

File: spec/models/org_space_policy_spec_with_annotations.rb
Comment (mkkeffeler) [reply spec/models/org_space_policy_spec_with_annotations.rb:77]: I am equally unsure. I would agree path has been taken but difference would be that the service broker api actually gave us ```organization_name``` and ```space_name``` values, right?

File: spec/models/org_space_policy_spec_with_annotations.rb
Comment (mkkeffeler) [reply spec/models/org_space_policy_spec_with_annotations.rb]: resolved with latest commit.

File: spec/models/org_space_policy_spec_with_annotations.rb
Comment (mkkeffeler) [reply spec/models/org_space_policy_spec_with_annotations.rb]: resolved with latest commit.

File: spec/models/org_space_policy_spec_no_annotations.rb
Comment (mkkeffeler) [reply spec/models/org_space_policy_spec_no_annotations.rb]: resolved with latest commit.

File: spec/models/org_space_policy_spec_no_annotations.rb
Comment (mkkeffeler) [reply spec/models/org_space_policy_spec_no_annotations.rb]: resolved with latest commit.

File: app/models/org_space_policy.rb
Comment (mkkeffeler) [reply app/models/org_space_policy.rb]: resolved with latest commit.

File: app/models/org_space_policy.rb
Comment (mkkeffeler) [reply app/models/org_space_policy.rb]: resolved with latest commit.

File: app/controllers/provision_controller.rb
Comment (mkkeffeler) [reply app/controllers/provision_controller.rb]: resolved with latest commit.

File: app/controllers/bind_controller.rb
Comment (mkkeffeler) [reply app/controllers/bind_controller.rb]: resolved with latest commit.

File: README.md
Comment (mkkeffeler) [reply README.md]: resolved with latest commit.

File: CHANGELOG.md
Comment (mkkeffeler) [reply CHANGELOG.md]: resolved with latest commit.

**Review by mkkeffeler — COMMENTED** [review]
> @mkkeffeler,
> 
> Thank you for submitting this PR! This looks very good!
> I've added some comments and suggestions.
> 
> One thing that might be good to add is an scenario/feature test that sets up values for the organization_name and space_name fields in the Service Broker API context. This would be VERY easy to add... it would be a minor variation of this test:
> https://github.com/cyberark/conjur-service-broker/blob/main/features/provision.feature#L48-L66
> 
> Also, for locally running unit tests as well as the feature/scenario tests, we have a pretty cool development environment. It's all menu-driven, VERY easy to use, and allows you to change code on the fly and re-test. It's described in the CONTRIBUTING.md here:
> https://github.com/cyberark/conjur-service-broker/blob/main/CONTRIBUTING.md#development-environment
> 
> As it says in the instructions, you just need to run 3 commands to get started:
> 
> ```
> ./dev/build
> ./dev/build_dev
> ./dev/start
> ```
> 
> and then select from the menu. The scripts will set up some local Conjur instances running in Docker-Compose, and allow you to test using those instances (esp. useful if you add a feature/scenario test).
> 
> Thanks again for the contribution!
> -Dane

Added as a feature!

**Comment by mkkeffeler** [comment]
@diverdane any updates?

**Comment by imheresamir** [comment]
Hi @mkkeffeler, it looks like commits 52e2b58 and 8ac3df4 need sign-offs amended to them. After that's done we can proceed to getting approval to building this PR in our pipeline. Thanks!

**Comment by rpothier** [comment]
Hi @mkkeffeler, could you squash the commits and rebase to the latest also?
Thanks

**Comment by mkkeffeler** [comment]
Done! @rpothier and @imheresamir

**Review by rpothier — CHANGES_REQUESTED** [review]
Thanks for squashing the commits, the rebase doesn't seem right though.
I think you are still missing a couple commits. When patching the changes I am getting
errors. Can you recheck it?

**Comment by mkkeffeler** [comment]
Changelog hates me, so it's gonna be 2 commits. Think I fixed rebase.

File: CHANGELOG.md
Comment (rpothier) [inline CHANGELOG.md:14]: Not a big deal, but if you need to make a change for another reason, please remove the extra line.

File: app/controllers/application_controller.rb
Comment (rpothier) [inline app/controllers/application_controller.rb:111]: Just a nit, Looks like an extra space, only if you have to make other changes.

File: spec/models/org_space_policy_spec_with_annotations.rb
Comment (mkkeffeler) [reply spec/models/org_space_policy_spec_with_annotations.rb:47]: good catch.

File: app/controllers/application_controller.rb
Comment (mkkeffeler) [reply app/controllers/application_controller.rb:111]: fixed

**Comment by imheresamir** [comment]
Please make sure the commits are signed-off so we can run a pipeline!

**Comment by mkkeffeler** [comment]
Done!

**Comment by imheresamir** [comment]
Looks like the changes corresponding to the recently resolved discussions haven't yet been included?

**Comment by mkkeffeler** [comment]
@imheresamir done! my bad.

File: CHANGELOG.md
Comment (rpothier) [reply CHANGELOG.md:14]: Actually not a nit, it causes a failure with the parse-changelog.sh script.
ERROR: line 14, column 1: Expected one of '### ', '## ', '[Unreleased]: ' at line 14, column 1 (byte 590) 
So we do need that line removed.

**Review by rpothier — CHANGES_REQUESTED** [review]
I'm seeing an error with the ./dev/test_integration test.
When I patched in your changes and ran it.
https://jenkins.conjur.net/blue/organizations/jenkins/cyberark--conjur-service-broker/detail/test-ci/3/pipeline/67

**Comment by mkkeffeler** [comment]
I tried viewing but said I don't have access

File: CHANGELOG.md
Comment (rpothier) [inline CHANGELOG.md:156]: Not sure why this like is changed, It must have a hidden character. Can you revert it?

File: features/provision.feature
Comment (rpothier) [inline features/provision.feature:68]: This test is failing. It is expecting 201, but getting 400. Also seeing the below error.
Then the HTTP response status code is "400"                                        # features/step_definitions/response_steps.rb:1

[2021-06-03T19:01:33.374Z]     And the JSON should be {}                                                          # json_spec-1.1.5/lib/json_spec/cucumber.rb:29

[2021-06-03T19:01:33.374Z]       Expected equivalent JSON

[2021-06-03T19:01:33.374Z]       Diff:

[2021-06-03T19:01:33.374Z]       @@ -1,295 +1,589 @@

[2021-06-03T19:01:33.374Z]        {

[2021-06-03T19:01:33.374Z]       +  "error": "Bad Request",

[2021-06-03T19:01:33.375Z]       +  "exception": "#<ActionDispatch::Http::Parameters::ParseError: 765: unexpected token at '{\n  \"context\": {\n    \"organization_guid\": \"e027f3f6-80fe-4d22-9374-da23a035ba0a\",\n    \"space_guid\": \"8c56f85c-c16e-4158-be79-5dac74f970db\"\n    \"organization_name\": \"my-organization\",\n    \"space_name\": \"my-space\"\n  },\n  \"service_id\": \"c024e536-6dc4-45c6-8a53-127e7f8275ab\",\n  \"plan_id\": \"3a116ac2-fc8b-496f-a715-e9a1b205d05c.community\",\n  \"organization_guid\": \"e027f3f6-80fe-4d22-9374-da23a035ba0a\",\n  \"space_guid\": \"8c56f85c-c16e-4158-be79-5dac74f970db\",\n  \"parameters\": {\n  }\n}'>",

[2021-06-03T19:01:33.375Z]       +  "status": 400,

**Review by rpothier — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (mkkeffeler) [reply CHANGELOG.md]: Change log looks good now @rpothier 

```
SC-MKEFFEL-OSX:conjur-service-broker mkeffele$ ./dev/parse-changelog.sh
++ dirname ./dev/parse-changelog.sh
+ cd ./dev
+ docker run --rm -v /Users/mkeffele/git_repos/conjur-service-broker/dev/..:/work -w /work ruby:2.5 bash -ec '
    gem install -N parse_a_changelog
    parse ./CHANGELOG.md
  '
Successfully installed polyglot-0.3.5
Successfully installed treetop-1.6.11
Successfully installed parse_a_changelog-1.1.0
3 gems installed
Parsing ./CHANGELOG.md
SUCCESS
```

**Review by mkkeffeler — COMMENTED** [review]
```SC-MKEFFEL-OSX:conjur-service-broker mkeffele$ ./dev/parse-changelog.sh
++ dirname ./dev/parse-changelog.sh
+ cd ./dev
+ docker run --rm -v /Users/mkeffele/git_repos/conjur-service-broker/dev/..:/work -w /work ruby:2.5 bash -ec '
    gem install -N parse_a_changelog
    parse ./CHANGELOG.md
  '
Successfully installed polyglot-0.3.5
Successfully installed treetop-1.6.11
Successfully installed parse_a_changelog-1.1.0
3 gems installed
Parsing ./CHANGELOG.md
SUCCESS```

File: features/provision.feature
Comment (mkkeffeler) [reply features/provision.feature:68]: Do we have a cause? not sure these logs tell us the issue with the code.

Comment (mkkeffeler) [reply features/provision.feature:68]: found it. was invalid JSON, no comma after ```space_guid```. fixed!

**Comment by rpothier** [comment]
Yeah! Everything is passing!

File: spec/models/org_space_policy_spec_with_annotations.rb
Comment (rpothier) [reply spec/models/org_space_policy_spec_with_annotations.rb:77]: So for this issue, the question remains if we could skip the next three tests. So does it make sense that we leave it in for now as it shouldn't be any worse?

**Comment by mkkeffeler** [comment]
Are we waiting on something from my side to get this merged? @rpothier

**Comment by rpothier** [comment]
There is one unresolved conversation. I commented on that. Are you ok with leaving it as is?

File: spec/models/org_space_policy_spec_with_annotations.rb
Comment (mkkeffeler) [reply spec/models/org_space_policy_spec_with_annotations.rb:77]: I think we leave this for now.

**Review by mkkeffeler — COMMENTED** [review]
> There is one unresolved conversation. I commented on that. Are you ok with leaving it as is?

Done.

**Review by rpothier — APPROVED** [review]

**Comment by rpothier** [comment]
Looks good! Thanks for your patience to get this through.

**Comment by mkkeffeler** [comment]
no thank you! when can we get a release cut with this feature? we are really wanting to test this out.

**Comment by izgeri** [comment]
@mkkeffeler are you installing the service broker via the tile or from the GitHub release of this project? We have a tile release scheduled to add support for TAS 2.11, but it's not scheduled right now until July.

**Comment by mkkeffeler** [comment]
via the tile at present. but can do the github release for short term if thats best.

**Comment by izgeri** [comment]
@mkkeffeler this has been released in v1.2.0 - there are install instructions for the service broker [here](https://github.com/cyberark/conjur-service-broker/tree/v1.2.0#installing-the-conjur-service-broker), but they are not super clear about upgrade v install. Upgrade is definitely much easier via the tile, but as I noted above I don't expect to release the tile until July.

**Comment by mkkeffeler** [comment]
Any rough ETA on the tile release?

**Comment by izgeri** [comment]
Tagging in @jtuttle and @alexkalish, who may have up-to-date information on this.

**Comment by mkkeffeler** [comment]
Thanks.

Needless to say this is a big blocker for us. We appreciate any help and as evidenced can contribute with certain things :)

**Comment by jtuttle** [comment]
@mkkeffeler Wanted to let you know what doing a tile release is on our radar and we'll be planning it into the upcoming sprint, so it should be released in the next few weeks. Thank you for your patience!

**Comment by mkkeffeler** [comment]
Great to hear.

How can we help support more frequent releases? This makes getting a change out very difficult, and can cause a lot of headache.

**Pull request merged**
